core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -342,17 +261,21 @@ class KafkaController(val config: KafkaConfig, zkUtils: ZkUtils, val brokerState
       partitionStateMachine.startup()
 
       // register the partition change listeners for all existing topics on failover
-      controllerContext.allTopics.foreach(topic => partitionStateMachine.registerPartitionChangeListener(topic))
+      controllerContext.allTopics.foreach(topic => registerPartitionModificationsListener(topic))
       info("Broker %d is ready to serve as the new controller with epoch %d".format(config.brokerId, epoch))
       maybeTriggerPartitionReassignment()
       maybeTriggerPreferredReplicaElection()
+      info("starting the controller scheduler")
+      kafkaScheduler.startup()
+      kafkaScheduler.schedule("controller-metric-task", () => controllerEventQueue.put(UpdateMetrics), period = 10, unit = TimeUnit.SECONDS)
```

Comment:
Do we need to update metrics in a scheduler? It seems that those metrics can only change after the processing of each event. If so, we can just update the metrics at the end of each event processing.